odoo/modules/registry.py

```diff
@@ -665,6 +665,12 @@ def check_indexes(self, cr, model_names):
                     expression = f'{column_expression} gin_trgm_ops'
                     method = 'gin'
                     where = ''
+                elif index == 'btree_not_null' and field.company_dependent:
+                    # company dependent condition will use extra
+                    # `AND col IS NOT NULL` to use the index.
+                    expression = f'({column_expression} IS NOT NULL)'
```

Comment:
Oh, so you only index the value `TRUE`, right?  That's definitively not what I would expect from such an index, although I understand the motivation.  IMHO this will be a endless source of confusion among developers.

Reply:
I takes much less disk
But I am not sure if we should introduce a new index. Or just use normal btree_not_null (i mean index the whole jsonb)

Reply:
What about one index per company? (Managed by the ORM)

Reply:
Yes, I am also trying to do that since current the search field_to_sql
`return SQL("COALESCE(%s->%s,to_jsonb(%s))::int4", sql_field, str(self.env.company.id), 0 if default is None else int(default))` cannot take advantage of the btree index, I want to refactor condition_to_sql to do that

Reply:
Yes it is used, and much faster if the column is sparse.
Also `USING btree ((data IS NOT NULL)) WHERE data IS NOT NULL;` takes smaller size than `USING btree (data) WHERE data IS NOT NULL;`

BTW an interesting result is condition like `(field_name->'1')::int = 100` can take advantage of both of the above index. It is really helpful if there is no fallback of a company dependent field
e.g if we define `(fields.Char(company_dependent=True, with_fallback=False)`
or we can remove the support for the fallback of the company dependent field as we discussed

And when we do the ondelete restrict logic e.g.
`WHERE field_name @? '$.* ? (@ == 200 || @ == 100)'`
it can also use the index